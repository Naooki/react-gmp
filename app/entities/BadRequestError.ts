export default interface BadRequestError {
  messages: string[];
}
