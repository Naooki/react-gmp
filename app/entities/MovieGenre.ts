enum MovieGenre {
  Crime = 'Crime',
  Documentary = 'Documentary',
  Horror = 'Horror',
  Drama = 'Drama',
  Romance = 'Romance',
  Animation = 'Animation',
  Adventure = 'Adventure',
  Family = 'Family',
  Comedy = 'Comedy',
  Fantasy = 'Fantasy',
  ScienceFiction = 'Science Fiction',
  Action = 'Action',
}

export default MovieGenre;
