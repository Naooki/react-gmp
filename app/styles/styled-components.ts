import * as styledComponents from 'styled-components';

// theme.ts
// your theme variables
export interface IThemeInterface {
  text: string;
  primary: string;
  componentBackground: string;
  componentBackgroundSecondary: string;
  footerBackground: string;
  pageBackground: string;
}

export const theme = {
  default: {
    text: '#fff',
    primary: '#f65261',
    componentBackground: '#232323',
    componentBackgroundSecondary: 'rgba(85,85,85,0.8)',
    footerBackground: '#424242',
    pageBackground: '#555',
  },
};
const {
  default: styled,
  css,
  createGlobalStyle,
  keyframes,
  ThemeProvider,
} = styledComponents as styledComponents.ThemedStyledComponentsModule<
  IThemeInterface
>;

export { css, createGlobalStyle, keyframes, ThemeProvider };
export default styled;
