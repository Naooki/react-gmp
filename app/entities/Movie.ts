import MovieGenre from './MovieGenre';

export interface Movie {
  id: string;
  releaseDate: string;
  about: string;
  title: string;
  imageUrl?: string;
  duration: number;
  rating: number;
  genres: MovieGenre[];
}
