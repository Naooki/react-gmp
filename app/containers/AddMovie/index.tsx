import * as React from 'react';

import { Movie } from 'entities/Movie';
import MovieForm from 'components/MovieForm';
import ModalHeading from 'containers/Modal/ModalHeading';
import { useDispatch } from 'react-redux';

interface Props {
  loading: boolean;
}
function AddMovie(props: Props) {
  const movie = {} as Movie;
  const dispatch = useDispatch();
  const onCreateMovieClick = React.useCallback(() => {
    dispatch(null);
  }, [dispatch]);

  return (
    <>
      <ModalHeading>add movie</ModalHeading>
      <MovieForm
        movie={movie}
        loading={props.loading}
        onConfirm={onCreateMovieClick}
      />
    </>
  );
}

export default AddMovie;
