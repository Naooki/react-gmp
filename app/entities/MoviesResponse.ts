import { Movie } from './Movie';

export default interface MoviesResponse {
  data: Movie[];
  total: number;
  offset: number;
  limit: number;
}
