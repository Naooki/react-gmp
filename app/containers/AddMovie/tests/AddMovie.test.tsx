import React from 'react';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components';
import { createStore, Store } from 'redux';

import { Movie } from 'entities/Movie';
import ActionTypes from 'containers/Movies/constants';
import moviesReducer from 'containers/Movies/reducer';
import { theme } from 'styles/styled-components';
import {
  act,
  cleanup,
  fireEvent,
  render,
  RenderResult,
} from '@testing-library/react';
import AddMovie from '..';

function renderComponent(): [Store, RenderResult] {
  const store = createStore(moviesReducer, {
    items: null,
    selectedItem: null,
  });
  store.dispatch = jest.fn();

  const utils = render(
    <Provider store={store}>
      <ThemeProvider theme={theme.default}>
        <AddMovie loading={false} />
      </ThemeProvider>
    </Provider>,
  );

  return [store, utils];
}

function populateTestData(
  renderResults: RenderResult,
  data?: Partial<Movie>,
): void {
  const title = renderResults.getByPlaceholderText(
    /Title here/,
  ) as HTMLInputElement;
  fireEvent.change(title, { target: { value: data?.title || 'Test title' } });

  const posterPath = renderResults.getByPlaceholderText(
    /Movie URL here/,
  ) as HTMLInputElement;
  fireEvent.change(posterPath, {
    target: { value: data?.poster_path || 'http://example.com/' },
  });

  const runtime = renderResults.getByPlaceholderText(
    /Runtime here/,
  ) as HTMLInputElement;
  fireEvent.change(runtime, {
    target: { value: data?.runtime?.toString() || '123' },
  });

  const overview = renderResults.getByPlaceholderText(
    /Overview here/,
  ) as HTMLInputElement;
  fireEvent.change(overview, {
    target: { value: data?.overview || 'Test overview' },
  });

  const genres = renderResults.getByPlaceholderText(
    /Select genre/,
  ) as HTMLInputElement;
  fireEvent.click(genres);

  const option = renderResults.getByText(/Horror/);
  fireEvent.click(option);
}

describe('<AddMovie />', () => {
  afterEach(cleanup);

  it('matches snapshot', () => {
    const [, renderResult] = renderComponent();
    expect(renderResult.asFragment()).toMatchSnapshot();
  });

  it('should dispatch createMovie action on submit', async () => {
    const [store, utils] = renderComponent();

    populateTestData(utils);

    await act(async () => {
      const submit = utils.getByText(/submit/);
      fireEvent.click(submit);
    });

    expect(store.dispatch).toHaveBeenCalledWith(
      expect.objectContaining({
        type: ActionTypes.CREATE_MOVIE,
        payload: {
          genres: ['Horror'],
          overview: 'Test overview',
          poster_path: 'http://example.com/',
          release_date: '',
          runtime: 123,
          title: 'Test title',
        },
      }),
    );
  });

  it('should not dispatch createMovie action when the the poster_path is invalid', async () => {
    const [store, utils] = renderComponent();

    populateTestData(utils, { poster_path: 'test-invalid-url' });

    await act(async () => {
      const submit = utils.getByText(/submit/);
      fireEvent.click(submit);
    });
    const uriError = utils.getByText(/Invalid URI/);

    expect(store.dispatch).not.toHaveBeenCalled();
    expect(uriError).toBeDefined();
  });
});
