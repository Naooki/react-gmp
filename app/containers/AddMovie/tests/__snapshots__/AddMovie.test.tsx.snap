// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddMovie /> matches snapshot 1`] = `
<DocumentFragment>
  <h1
    class="sc-fzoLsD sc-fzpans pEcfq"
  >
    add movie
  </h1>
  <form
    action="#"
    class="sc-fzozJi fuiiNa"
  >
    <div
      class="sc-AxirZ ebjXWE"
    >
      <label
        class="sc-AxiKw ildPaq"
        for="title"
      >
        title*
      </label>
      <input
        class="sc-AxhUy XSPAy"
        id="title"
        name="title"
        placeholder="Title here"
        type="text"
        value=""
      />
    </div>
    <div
      class="sc-AxirZ ebjXWE"
    >
      <label
        class="sc-AxiKw ildPaq"
        for="release_date"
      >
        release date
      </label>
      <input
        class="sc-AxhUy XSPAy"
        id="release_date"
        name="release_date"
        placeholder="Select date"
        type="date"
        value=""
      />
    </div>
    <div
      class="sc-AxirZ ebjXWE"
    >
      <label
        class="sc-AxiKw ildPaq"
        for="poster_path"
      >
        movie url*
      </label>
      <input
        class="sc-AxhUy XSPAy"
        id="poster_path"
        name="poster_path"
        placeholder="Movie URL here"
        type="text"
        value=""
      />
    </div>
    <div
      class="sc-AxirZ ebjXWE"
    >
      <label
        class="sc-AxiKw ildPaq"
        for="genres"
      >
        genre*
      </label>
      <input
        class="sc-AxhUy XSPAy"
        id="genres"
        placeholder="Select genre"
        readonly=""
        value=""
      />
      <div
        class="sc-AxgMl ecWowd"
      />
    </div>
    <div
      class="sc-AxirZ ebjXWE"
    >
      <label
        class="sc-AxiKw ildPaq"
        for="overview"
      >
        overview*
      </label>
      <input
        class="sc-AxhUy XSPAy"
        id="overview"
        name="overview"
        placeholder="Overview here"
        type="text"
        value=""
      />
    </div>
    <div
      class="sc-AxirZ ebjXWE"
    >
      <label
        class="sc-AxiKw ildPaq"
        for="runtime"
      >
        runtime*
      </label>
      <input
        class="sc-AxhUy XSPAy"
        id="runtime"
        name="runtime"
        placeholder="Runtime here"
        type="number"
        value=""
      />
    </div>
    <div
      class="sc-AxjAm qXCZf"
    >
      <button
        class="sc-AxmLO vUKCa disabled"
        disabled=""
        type="reset"
      >
        reset
      </button>
      <button
        class="sc-AxmLO vUKCa contained"
        type="submit"
      >
        submit
      </button>
    </div>
  </form>
</DocumentFragment>
`;
